File: leetcode/two pointer/986. Interval List Intersections/js/main.js
// Time: O(M + N), where M, N are the lengths of nums1 and nums2.
// Space: O(N), the maximum size of the answer.
const intervalIntersection = (nums1, nums2) => {
  let i = 0;
  let j = 0;
  let ans = [];

  while (i < nums1.length && j < nums2.length) {
    let first = Math.max(nums1[i][0], nums2[j][0]);
    let second = Math.min(nums1[i][1], nums2[j][1]);

    if (first <= second) ans.push([first, second]);

    if (nums1[i][1] < nums2[j][1]) {
      i++;
    } else {
      j++;
    }
  }

  return ans;
};

File: leetcode/two pointer/986. Interval List Intersections/py/main.py
class Solution:
    # Time O(N + M) where M, NM,N are the lengths of A and B respectively.
    # Space O(N + M)
    def intervalIntersection(self, A: List[List[int]],
                             B: List[List[int]]) -> List[List[int]]:
        i, j = 0, 0
        ans = []

        while i < len(A) and j < len(B):
            a_start, a_end = A[i]
            b_start, b_end = B[j]

            start, end = max(a_start, b_start), min(a_end, b_end)

            if start <= end:
                ans.append([start, end])

            if a_end < b_end:
                i += 1
            else:
                j += 1

        return ans
